tsweb/tsweb.go

```diff
@@ -447,6 +449,55 @@ func VarzHandler(w http.ResponseWriter, r *http.Request) {
 	varz.Handler(w, r)
 }
 
+// CleanRedirectURL ensures that urlStr is a valid redirect URL to the
+// current server, or one of allowedHosts. Returns the cleaned URL or
+// a validation error.
+func CleanRedirectURL(urlStr string, allowedHosts []string) (*url.URL, error) {
+	// In some places, we unfortunately query-escape the redirect URL
+	// too many times, and end up needing to redirect to a URL that's
+	// still escaped by one level. Try to unescape the input.
+	unescaped, err := url.QueryUnescape(urlStr)
+	if err == nil && unescaped != urlStr {
+		urlStr = unescaped
+	}
+
+	// Go's URL parser and browser URL parsers disagree on the meaning
+	// of malformed HTTP URLs. Given the input https:/evil.com, Go
+	// parses it as hostname="", path="/evil.com". Browsers parse it
+	// as hostname="evil.com", path="". This means that, using
+	// malformed URLs, an attacker could trick us into approving of a
+	// "local" redirect that in fact sends people elsewhere.
+	//
+	// This very blunt check enforces that we'll only process
+	// redirects that are definitely well-formed URLs.
+	//
+	// Note that the check for just / also allows URLs of the form
+	// "//foo.com/bar", which are scheme-relative redirects. These
+	// must be handled with care below when determining whether a
+	// redirect is relative to the current host. Notably,
+	// url.URL.IsAbs reports // URLs as relative, whereas we want to
+	// treat them as absolute redirects and verify the target host.
+	if !strings.HasPrefix(urlStr, "/") && !strings.HasPrefix(urlStr, "http://") && !strings.HasPrefix(urlStr, "https://") {
```

Comment:
Case-insensitive compare?

Reply:
Hmm, there's no handy strings funcs for that, and we fail safe by rejecting redirects that play silly games with scheme casing. Empirically we've not needed the extra flexibility so far, so I'm inclined to keep this validator more strict? WDYT?

Reply:
Well, I'd hate for someone to find out they can bypass the fix just by saying "HTTPS:/evil.com". The standard library IMO should offer `strings.HasPrefixFunc` (and, for that matter, `strings.ContainsFunc`) and we could pass `strings.EqualFold` to it. But, we can make our own, and put it in tailscale/util?

```
func hasPrefixEqualFold(haystack []string, needle string) bool {
	for _, s := range haystack {
		if strings.EqualFold(s, needle) {
			return true
		}
	}
	return false
}
```

Reply:
Per discussion in slack: updated to do case folding, on both the prefix and hostname tests